### PR TITLE
refactor: Migrates Datasource: mongodbatlas_advanced_cluster & mongodbatlas_advanced_clusters to Terraform Plugin Framework

### PR DESCRIPTION
**Pull request by maastha**
## Description

Please include a summary of the fix/feature/change, including any relevant motivation and context.

Link to any related issue(s):

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR.
- [ ] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [ ] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [ ] I have read the [contribution guidelines](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/CONTRIBUTING.md)
- [ ] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [ ] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [ ] I have added any necessary documentation (if appropriate)
- [ ] I have run make fmt and formatted my code
- [ ] If changes include deprecations or removals, I defined an isolated PR with a relevant title as it will be used in the auto-generated changelog.

## Comments
Key changes observed after migrating:

- collection order change for `region_configs` (region_configs is a TypeSet so order cannot be guaranteed)
- Attributes with empty `"", false, 0` updated to `null`

## Testing: 
Steps:
1. Create resource mongodbatlas_advanced_cluster with given Terraform version and mongodbatlas provider version
2. Run `terraform plan` and `terraform apply` to write data source output to the state file
3. Override mongodbatlas provider version to local (dev overrides)
4. Run `terraform plan` and `terraform apply` to review plan and write data source output to the state file
5. Compare final state outputs via text diff

Config for outputs:
```
# Output the entire cluster configuration
output "complete_cluster_info_global" {
  value = data.mongodbatlas_advanced_cluster.global-adv-cluster
}

output "complete_cluster_info_tenant" {
  value = data.mongodbatlas_advanced_cluster.example
}

output "all_clusters_in_project_info" {
  value = data.mongodbatlas_advanced_clusters.clusters-ds
}
```

**Terraform version 1.0.8**

<details>

_mongodbatlas provider version: 1.11.0_:

data source: mongodbatlas_advanced_cluster
https://www.diffchecker.com/5ngOj7rA/

data source: mongodbatlas_advanced_clusters
https://www.diffchecker.com/KQC75AFC/

_mongodbatlas provider version: 1.13.1_:

data source: mongodbatlas_advanced_cluster
https://www.diffchecker.com/MQBgF1df/
image

data source: mongodbatlas_advanced_clusters
https://www.diffchecker.com/hbKRx846/

</details>

**Terraform version 1.6.5**
<details>

_mongodbatlas provider version: 1.11.0_:
data source: mongodbatlas_advanced_cluster
https://www.diffchecker.com/RxenoIKn/
image

data source: mongodbatlas_advanced_clusters
https://www.diffchecker.com/RPUd8JX9/

_mongodbatlas provider version: 1.13.1_:

data source: mongodbatlas_advanced_cluster
https://www.diffchecker.com/yPfXVJdE/

data source: mongodbatlas_advanced_clusters
https://www.diffchecker.com/w5B3p39J/

</details>

### CODE REVIEW

File: internal/service/advancedcluster/fw_data_source_advanced_cluster.go [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func advancedClusterRegionConfigAutoScalingSpecsAttr() schema.ListNestedAttribute {
```
Comment (andreaangiolillo) [inline internal/service/advancedcluster/fw_data_source_advanced_cluster.go]: advance or adv? Let's define one of these and use it everywhere
```suggestion
func advClusterRegionConfigAutoScalingSpecsAttr() schema.ListNestedAttribute {
```

**Review by andreaangiolillo — APPROVED** [review]
LGTM

File: internal/common/conversion/type_conversion.go [hunk 2]
```diff
@@ -56,3 +59,10 @@ func IsStringPresent(strPtr *string) bool {
 func MongoDBRegionToAWSRegion(region string) string {
 	return strings.ReplaceAll(strings.ToLower(region), "_", "-")
 }
+
+func FloatToTfInt64(v *float64) types.Int64 {
```
Comment (lantoli) [inline internal/common/conversion/type_conversion.go:63]: do we want to handle conversion errors?

**Review by lantoli — DISMISSED** [review]
we're not using fw_ any more in filenames

File: internal/service/advancedcluster/fw_data_source_advanced_cluster.go [hunk 3]
```diff
@@ -0,0 +1,429 @@
+package advancedcluster
```
Comment (maastha) [inline internal/service/advancedcluster/fw_data_source_advanced_cluster.go:1]: This file will be renamed to `data_source_advanced_cluster.go` as per naming convention, currently using `fw` so review is easier

File: internal/service/advancedcluster/fw_data_source_advanced_clusters.go [hunk 4]
```diff
@@ -0,0 +1,183 @@
+package advancedcluster
```
Comment (maastha) [inline internal/service/advancedcluster/fw_data_source_advanced_clusters.go:1]: This file will be renamed to `data_source_advanced_clusters.go` as per naming convention, currently using `fw` so review is easier

**Comment by maastha** [comment]
> we're not using fw_ any more in filenames

Hi @lantoli 
My bad I've added comments to those files, yes I will be renaming the files as per convention before merge, currently using `fw` prefix as the PR was getting super noisy for these large files once I rename them using the current names

File: internal/service/advancedcluster/fw_data_source_advanced_cluster.go [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+			}
+		}
+	}
+
+	tfContainersIDsMap, diags = types.MapValue(types.StringType, containerIDsMap)
+
+	return tfRegionConfigs, tfContainersIDsMap, diags
+}
+
+func advancedClusterDSAttributes() map[string]schema.Attribute {
```
Comment (AgustinBettati) [inline internal/service/advancedcluster/fw_data_source_advanced_cluster.go:198]: would keep the schema definition at the top just for convetion

File: internal/service/advancedcluster/model_advanced_cluster.go [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+	"oplog_min_retention_hours":            types.Int64Type,
+	"sample_size_bi_connector":             types.Int64Type,
+	"sample_refresh_interval_bi_connector": types.Int64Type,
+	"transaction_lifetime_limit_seconds":   types.Int64Type,
+	"fail_index_key_too_long":              types.BoolType,
+	"javascript_enabled":                   types.BoolType,
+	"no_table_scan":                        types.BoolType,
+}}
+
+type tfConnectionStringModel struct {
```
Comment (AgustinBettati) [inline internal/service/advancedcluster/model_advanced_cluster.go:73]: some models/types are public (upercase) while others are private (lowercase). Would unify, usually we are keeping them private but when adding unit testing the have to be made public.

File: internal/service/advancedcluster/fw_data_source_advanced_clusters.go [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+				DeprecationMessage: "Please use each cluster's id attribute instead",
+				Computed:           true,
+			},
+			"project_id": schema.StringAttribute{
+				Required: true,
+			},
+			"results": schema.ListNestedAttribute{
+				Computed: true,
+				NestedObject: schema.NestedAttributeObject{
+					Attributes: advancedClusterDSAttributes(),
```
Comment (AgustinBettati) [inline internal/service/advancedcluster/fw_data_source_advanced_clusters.go:52]: Not certain if there is a risk here: `advancedClusterDSAttributes()` defines `project_id` as required, which would not be correct in the case of computed results for this data source. As an example, in stream instance a boolean is passed to handle this difference but be able to reuse the schema: https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/internal/service/streaminstance/data_source_stream_instance.go#L35

File: internal/service/advancedcluster/model_advanced_cluster.go [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+	matlas "go.mongodb.org/atlas/mongodbatlas"
+
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+
+	"github.com/mongodb/terraform-provider-mongodbatlas/internal/common/conversion"
+)
+
+type TfLabelModel struct {
```
Comment (AgustinBettati) [inline internal/service/advancedcluster/model_advanced_cluster.go:15]: just out of convention from other resource would keep terraform related models in same file were schema is defined.

**Review by AgustinBettati — APPROVED** [review]
LGTM, comments are non-blocking

File: internal/service/advancedcluster/model_advanced_cluster.go (hunk 8 above)
Comment (maastha) [reply internal/service/advancedcluster/model_advanced_cluster.go:15]: so the idea of model_*.go files is just to have conversion methods?

**Comment by marcosuma** [comment]
@maastha thank you for your work! Two comments:
1. let's wait to go with 1.14.0 before merging this
2. if we merge it without merging resource migration right after and let's say there is an urgency and we have to release 1.14.1, will there be unexpected behaviour if DS is using TPF and Resource is using SDKv2?

File: internal/service/advancedcluster/model_advanced_cluster.go (hunk 8 above)
Comment (AgustinBettati) [reply internal/service/advancedcluster/model_advanced_cluster.go:15]: yes, and model_*_test.go for testing those conversion methods. I imagine we will likely have schema and terraform model definitions also in a separate file when we starting autogenerating them from the api spec.

**Pull request closed without merging**